File: LeetCode/DesignLinkedList.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


class MyLinkedList:

    def __init__(self):
        self.head = None
        self.tail = None

    def get(self, index):
        current = self.head
        idx = 0

        while current:
            if idx == index:
                return current.val

            current = current.next
            idx += 1

        return -1

    def addAtHead(self, val):
        new_node = Node(val)

        if self.head is None:
            self.head = new_node
            self.tail = new_node
            return

        new_node.next = self.head
        self.head = new_node

    def addAtTail(self, val):
        new_node = Node(val)

        if self.head is None:
            self.head = new_node
            self.tail = new_node
            return

        self.tail.next = new_node
        self.tail = new_node

    def addAtIndex(self, index, val):
        if index < 0:
            return

        # Find current length
        length = 0
        current = self.head

        while current:
            length += 1
            current = current.next

        # Invalid index
        if index > length:
            return

        # Insert at beginning
        if index == 0:
            self.addAtHead(val)
            return

        # Insert at end
        if index == length:
            self.addAtTail(val)
            return

        # Insert in middle
        current = self.head
        idx = 0

        while idx < index - 1:
            current = current.next
            idx += 1

        new_node = Node(val)

        new_node.next = current.next
        current.next = new_node

    def deleteAtIndex(self, index):
        if self.head is None or index < 0:
            return

        if index == 0:
            self.head = self.head.next
            if self.head is None:
                self.tail = None
            return

        current = self.head
        prev = None
        idx = 0

        while current and idx < index:
            prev = current
            current = current.next
            idx += 1

        if current is None:
            return

        prev.next = current.next

        if current is self.tail:
            self.tail = prev


# # Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)

File: BasicToAdvance/Java/LeetCode/DesignLinkedList.java
public class DesignLinkedList {
	private static class Node {
		int val;
		Node next;

		Node(int val) {
			this.val = val;
		}
	}

	public static class MyLinkedList {
		private Node head;
		private Node tail;

		public int get(int index) {
			Node current = head;
			for (int i = 0; current != null; i++, current = current.next) {
				if (i == index) {
					return current.val;
				}
			}
			return -1;
		}

		public void addAtHead(int val) {
			Node node = new Node(val);
			node.next = head;
			head = node;
			if (tail == null) {
				tail = node;
			}
		}

		public void addAtTail(int val) {
			Node node = new Node(val);
			if (tail == null) {
				head = tail = node;
				return;
			}
			tail.next = node;
			tail = node;
		}

		public void addAtIndex(int index, int val) {
			if (index < 0) {
				return;
			}

			int length = 0;
			Node current = head;
			while (current != null) {
				length++;
				current = current.next;
			}

			if (index > length) {
				return;
			}
			if (index == 0) {
				addAtHead(val);
				return;
			}
			if (index == length) {
				addAtTail(val);
				return;
			}

			current = head;
			for (int i = 1; i < index; i++) {
				current = current.next;
			}
			Node node = new Node(val);
			node.next = current.next;
			current.next = node;
		}

		public void deleteAtIndex(int index) {
			if (index < 0 || head == null) {
				return;
			}
			if (index == 0) {
				head = head.next;
				if (head == null) {
					tail = null;
				}
				return;
			}

			Node previous = head;
			for (int i = 1; i < index && previous.next != null; i++) {
				previous = previous.next;
			}
			if (previous.next == null) {
				return;
			}
			if (previous.next == tail) {
				tail = previous;
			}
			previous.next = previous.next.next;
		}
	}
}
